File: .claude/hooks/validate-tiknix-php.py
#!/usr/bin/env python3
"""
Tiknix PHP Code Validator Hook

Validates PHP code against Tiknix/RedBeanPHP/FlightPHP coding standards:
1. Bean property names should use camelCase (not underscore_case)
2. Table names should use camelCase (not underscore_case)
3. R::exec should almost NEVER be used - only in extreme situations
"""

import json
import sys
import re


def find_underscore_properties(content):
    """
    Find bean properties using underscore_case instead of camelCase.

    Note: RedBeanPHP accepts both snake_case and camelCase for properties.
    We PREFER camelCase but don't block on snake_case since both work.
    This returns an empty list - snake_case is allowed but camelCase is preferred.
    """
    # Both conventions work in RedBeanPHP, so we don't block on this
    # Preference for camelCase is documented in CLAUDE.md
    return []


def find_underscore_table_names(content):
    """
    Find R::dispense with invalid bean type names - these WILL FAIL at runtime!

    CRITICAL: R::dispense() bean type names must be:
    - All lowercase (a-z)
    - Only alphanumeric (no underscores, no uppercase)

    The error will be: "Invalid bean type: table_name"

    For R::findOne, R::find, R::load - the bean type is more flexible
    because RedBeanPHP converts it. But R::dispense is strict.

    Must use all lowercase for R::dispense (e.g., 'enterprisesettings' not 'enterprise_settings' or 'enterpriseSettings')
    """
    issues = []

    # Match R::dispense with any table name
    pattern = r"R::dispense\s*\(\s*['\"]([a-zA-Z0-9_]+)['\"]"

    for match in re.finditer(pattern, content):
        table_name = match.group(1)

        # Check for underscores
        if '_' in table_name:
            # Convert to suggested lowercase
            lowercase = table_name.replace('_', '').lower()
            issues.append(
                f"R::dispense('{table_name}') will FAIL! RedBeanPHP doesn't allow underscores in dispense(). "
                f"Use R::dispense('{lowercase}') instead."
            )
        # Check for uppercase letters
        elif table_name != table_name.lower():
            lowercase = table_name.lower()
            issues.append(
                f"R::dispense('{table_name}') will FAIL! RedBeanPHP requires all lowercase bean types in dispense(). "
                f"Use R::dispense('{lowercase}') instead."
            )

    return issues


def find_exec_usage(content):
    """Find problematic use of R::exec and flag it for review."""
    issues = []

    # Match R::exec with any SQL statement
    pattern = r"R::exec\s*\(\s*['\"]([^'\"]+)['\"]"

    for match in re.finditer(pattern, content):
        sql = match.group(1).strip()
        sql_upper = sql.upper()

        # DDL operations are OK - these can't be done with beans
        if sql_upper.startswith('CREATE ') or sql_upper.startswith('ALTER ') or sql_upper.startswith('DROP '):
            continue  # DDL is acceptable

        # Check what type of operation it is
        if sql_upper.startswith('INSERT'):
            issues.append(f"R::exec() used for INSERT. This bypasses FUSE models! Use Bean::dispense() + Bean::store() instead.")
        elif sql_upper.startswith('UPDATE'):
            # Check if it's a simple update that should use beans
            if 'WHERE' in sql_upper and ('= ?' in sql or '=?' in sql):
                # Check if it's NOT a complex operation (increment, bulk, etc.)
                if '+ 1' not in sql and '- 1' not in sql and 'NOW()' not in sql_upper:
                    issues.append(f"R::exec() used for UPDATE. This bypasses FUSE models! Use Bean::load() + Bean::store() instead.")
                else:
                    issues.append(f"R::exec() for UPDATE detected. Verify this is truly necessary and cannot be done with beans.")
            else:
                issues.append(f"R::exec() for UPDATE detected. Verify this is truly necessary and cannot be done with beans.")
        elif sql_upper.startswith('DELETE'):
            issues.append(f"R::exec() used for DELETE. This bypasses FUSE models! Use Bean::trash() instead.")
        else:
            issues.append(f"R::exec() detected. R::exec should ONLY be used in extreme situations. Can this use bean methods instead?")

    return issues


def validate_php_code(content):
    """Run all validations on PHP content."""
    all_issues = []

    # Skip if not PHP
    if '<?php' not in content and '<?=' not in content:
        # Check if it contains PHP-like RedBean code even without <?php tag
        if 'R::' not in content:
            return []

    all_issues.extend(find_underscore_properties(content))
    all_issues.extend(find_underscore_table_names(content))
    all_issues.extend(find_exec_usage(content))

    return all_issues


def main():
    try:
        # Read input from stdin (JSON format from Claude Code)
        input_data = json.load(sys.stdin)

        tool_name = input_data.get('tool_name', '')
        tool_input = input_data.get('tool_input', {})

        # Only validate Write and Edit operations
        if tool_name not in ['Write', 'Edit']:
            sys.exit(0)

        # Get file path and content
        file_path = tool_input.get('file_path', '')

        # Only validate PHP files
        if not file_path.endswith('.php'):
            sys.exit(0)

        # Get the content being written/edited
        if tool_name == 'Write':
            content = tool_input.get('content', '')
        elif tool_name == 'Edit':
            content = tool_input.get('new_string', '')
        else:
            sys.exit(0)

        # Run validations
        issues = validate_php_code(content)

        if issues:
            # Format feedback message
            feedback = "TIKNIX CODE STANDARDS VIOLATION:\n\n"
            for i, issue in enumerate(issues, 1):
                feedback += f"{i}. {issue}\n"
            feedback += "\nSee CLAUDE.md for Tiknix coding standards.\n"
            feedback += "RedBeanPHP docs: https://redbeanphp.com/"

            # Return JSON with block decision
            output = {
                "decision": "block",
                "reason": feedback
            }
            print(json.dumps(output))

        sys.exit(0)

    except json.JSONDecodeError:
        # If input isn't valid JSON, just pass through
        sys.exit(0)
    except Exception as e:
        # Log error but don't block
        print(f"Hook error: {e}", file=sys.stderr)
        sys.exit(0)


if __name__ == '__main__':
    main()

File: .claude/hooks/validate-tiknix-php.php
#!/usr/bin/env php
<?php
/**
 * Tiknix PHP Code Validator Hook
 *
 * Validates PHP code against Tiknix/RedBeanPHP/FlightPHP coding standards:
 * 1. Bean type names must be all lowercase (no underscores) for R::dispense
 * 2. R::exec should almost NEVER be used - only in extreme situations
 * 3. Prefer RedBeanPHP associations (ownBeanList/sharedBeanList) over manual FK management
 * 4. Controller class names must be all lowercase - NO CamelCase, NO hyphens
 * 5. Route files must not contain hyphens
 */

/**
 * Find R::dispense with invalid bean type names - these WILL FAIL at runtime!
 */
function findUnderscoreTableNames(string $content): array {
    $issues = [];

    // Match R::dispense with any table name
    if (preg_match_all("/R::dispense\s*\(\s*['\"]([a-zA-Z0-9_]+)['\"]/", $content, $matches)) {
        foreach ($matches[1] as $tableName) {
            // Check for underscores
            if (strpos($tableName, '_') !== false) {
                $lowercase = strtolower(str_replace('_', '', $tableName));
                $issues[] = "R::dispense('{$tableName}') will FAIL! RedBeanPHP doesn't allow underscores in dispense(). Use R::dispense('{$lowercase}') instead.";
            }
            // Check for uppercase letters
            elseif ($tableName !== strtolower($tableName)) {
                $lowercase = strtolower($tableName);
                $issues[] = "R::dispense('{$tableName}') will FAIL! RedBeanPHP requires all lowercase bean types in dispense(). Use R::dispense('{$lowercase}') instead.";
            }
        }
    }

    return $issues;
}

/**
 * Find problematic use of R::exec and flag it for review.
 */
function findExecUsage(string $content): array {
    $issues = [];

    if (preg_match_all("/R::exec\s*\(\s*['\"]([^'\"]+)['\"]/", $content, $matches)) {
        foreach ($matches[1] as $sql) {
            $sqlUpper = strtoupper(trim($sql));

            // DDL operations are OK
            if (preg_match('/^(CREATE|ALTER|DROP)\s/', $sqlUpper)) {
                continue;
            }

            if (strpos($sqlUpper, 'INSERT') === 0) {
                $issues[] = "R::exec() used for INSERT. This bypasses FUSE models! Use Bean::dispense() + Bean::store() instead.";
            } elseif (strpos($sqlUpper, 'UPDATE') === 0) {
                $issues[] = "R::exec() for UPDATE detected. This bypasses FUSE models! Use Bean::load() + Bean::store() instead.";
            } elseif (strpos($sqlUpper, 'DELETE') === 0) {
                $issues[] = "R::exec() used for DELETE. This bypasses FUSE models! Use Bean::trash() instead.";
            } else {
                $issues[] = "R::exec() detected. R::exec should ONLY be used in extreme situations. Can this use bean methods instead?";
            }
        }
    }

    return $issues;
}

/**
 * Detect manual foreign key assignments and suggest using associations instead.
 */
function findManualFkAssignments(string $content): array {
    $issues = [];

    // Common FK column patterns
    $knownFks = ['board_id', 'job_id', 'repo_id', 'member_id', 'parent_id'];

    // Match $bean->xxx_id =
    if (preg_match('/\$\w+->\w+_id\s*=/', $content)) {
        $issues[] = "Manual FK assignment detected (->xxx_id =). Consider using RedBeanPHP associations: \$parent->ownChildList[] = \$child (auto-sets FK, lazy loading). See CLAUDE.md for examples.";
    }

    return $issues;
}

/**
 * Check controller naming - BLOCKS CamelCase and hyphens
 * Returns: [blockingIssues, warningIssues]
 */
function checkControllerNaming(string $filePath): array {
    $blocking = [];
    $warnings = [];

    // Only check files in controls/ directory
    if (strpos($filePath, '/controls/') === false) {
        return [[], []];
    }

    // Skip BaseControls subdirectory
    if (strpos($filePath, 'BaseControls') !== false) {
        return [[], []];
    }

    $filename = basename($filePath);
    if (!str_ends_with($filename, '.php')) {
        return [[], []];
    }

    $controllerName = substr($filename, 0, -4); // Remove .php

    // Count capital letters (excluding first letter)
    $capitalCount = 0;
    for ($i = 1; $i < strlen($controllerName); $i++) {
        if (ctype_upper($controllerName[$i])) {
            $capitalCount++;
        }
    }

    // BLOCKING: CamelCase controllers
    if ($capitalCount > 0) {
        $lowercase = strtolower($controllerName);
        $blocking[] = "CamelCase controller '{$controllerName}' is FORBIDDEN. " .
            "FlightPHP auto-routing requires all lowercase controller names. " .
            "Rename to '{$lowercase}' (file: {$lowercase}.php, class: {$lowercase}).";
    }

    return [$blocking, $warnings];
}

/**
 * Check route file naming - BLOCKS hyphens
 */
function checkRouteFileNaming(string $filePath): array {
    $blocking = [];

    // Only check files in routes/ directory
    if (strpos($filePath, '/routes/') === false) {
        return [];
    }

    $filename = basename($filePath);

    // BLOCKING: Hyphens in route file names
    if (strpos($filename, '-') !== false) {
        $fixed = str_replace('-', '', $filename);
        $blocking[] = "Hyphenated route file '{$filename}' is FORBIDDEN. " .
            "Hyphens in URLs require explicit route files, which defeats auto-routing. " .
            "Rename controller and route file to remove hyphens: '{$fixed}'.";
    }

    return $blocking;
}

/**
 * Run all validations on PHP content.
 */
function validatePhpCode(string $content, string $filePath): array {
    $blockingIssues = [];
    $warningIssues = [];

    // File path based checks (always run)
    [$controllerBlocking, $controllerWarnings] = checkControllerNaming($filePath);
    $blockingIssues = array_merge($blockingIssues, $controllerBlocking);
    $warningIssues = array_merge($warningIssues, $controllerWarnings);

    $blockingIssues = array_merge($blockingIssues, checkRouteFileNaming($filePath));

    // Skip content checks if not PHP
    if (strpos($content, '<?php') === false && strpos($content, '<?=') === false) {
        if (strpos($content, 'R::') === false && strpos($content, 'Bean::') === false) {
            return [$blockingIssues, $warningIssues];
        }
    }

    // Blocking issues - these will cause runtime errors
    $blockingIssues = array_merge($blockingIssues, findUnderscoreTableNames($content));

    // Warning issues - suggestions for better practices
    $warningIssues = array_merge($warningIssues, findExecUsage($content));
    $warningIssues = array_merge($warningIssues, findManualFkAssignments($content));

    return [$blockingIssues, $warningIssues];
}

// Main execution
$input = file_get_contents('php://stdin');
$data = json_decode($input, true);

if (!$data) {
    exit(0);
}

$toolName = $data['tool_name'] ?? '';
$toolInput = $data['tool_input'] ?? [];

// Only validate Write and Edit operations
if (!in_array($toolName, ['Write', 'Edit'])) {
    exit(0);
}

$filePath = $toolInput['file_path'] ?? '';

// Only validate PHP files
if (!str_ends_with($filePath, '.php')) {
    exit(0);
}

// Get content being written/edited
$content = '';
if ($toolName === 'Write') {
    $content = $toolInput['content'] ?? '';
} elseif ($toolName === 'Edit') {
    $content = $toolInput['new_string'] ?? '';
}

// Run validations
[$blockingIssues, $warningIssues] = validatePhpCode($content, $filePath);

// Blocking issues - will prevent the operation
if (!empty($blockingIssues)) {
    $feedback = "TIKNIX CODE STANDARDS VIOLATION (BLOCKING):\n\n";
    foreach ($blockingIssues as $i => $issue) {
        $feedback .= ($i + 1) . ". {$issue}\n";
    }
    $feedback .= "\nThese issues will cause runtime errors. Fix before proceeding.\n";
    $feedback .= "See CLAUDE.md for Tiknix coding standards.";

    echo json_encode([
        'decision' => 'block',
        'reason' => $feedback
    ]);
    exit(0);
}

// Warning issues - allow but inform
if (!empty($warningIssues)) {
    $feedback = "TIKNIX BEST PRACTICES SUGGESTION:\n\n";
    foreach ($warningIssues as $i => $issue) {
        $feedback .= ($i + 1) . ". {$issue}\n";
    }
    $feedback .= "\nThese are suggestions for better code. Operation allowed.\n";
    $feedback .= "See CLAUDE.md for patterns.";

    echo json_encode([
        'decision' => 'allow',
        'reason' => $feedback
    ]);
}

exit(0);
